File: hammer-gui/src/main/java/dev/vini2003/hammer/gui/api/client/screen/base/BaseHandledScreen.java
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.gui.api.client.screen.base;

import dev.vini2003.hammer.core.api.client.util.InstanceUtil;
import dev.vini2003.hammer.core.api.client.util.PositionUtil;
import dev.vini2003.hammer.gui.api.common.event.*;
import dev.vini2003.hammer.gui.api.common.screen.handler.BaseScreenHandler;
import dev.vini2003.hammer.gui.api.common.widget.Widget;
import dev.vini2003.hammer.gui.registry.common.HGUINetworking;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.minecraft.client.gui.screen.ingame.HandledScreen;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.text.Text;

import java.util.List;

public abstract class BaseHandledScreen<T extends BaseScreenHandler> extends HandledScreen<T> {
	public BaseHandledScreen(T screenHandler, PlayerInventory playerInventory, Text text) {
		super(screenHandler, playerInventory, text);
	}
	
	@Override
	protected void init() {
		handler.getChildren().clear();
		handler.getSlots().clear();
		
		backgroundWidth = width;
		backgroundHeight = height;
		
		super.init();
		
		handler.init(width, height);
		
		handler.onLayoutChanged();
		
		for (var child : handler.getAllChildren()) {
			child.dispatchEvent(new LayoutChangedEvent());
		}
		
		var buf = PacketByteBufs.create();
		buf.writeInt(width);
		buf.writeInt(height);
		
		ClientPlayNetworking.send(HGUINetworking.SYNC_SCREEN_HANDLER, buf);
	}
	
	@Override
	protected void drawBackground(MatrixStack matrices, float delta, int mouseX, int mouseY) {}
	
	@Override
	protected boolean isClickOutsideBounds(double mouseX, double mouseY, int left, int top, int button) {
		for (var child : handler.getAllChildren()) {
			if (child.isPointWithin((float) mouseX, (float) mouseY)) {
				return false;
			}
		}
		
		return true;
	}
	
	@Override
	public boolean mouseClicked(double mouseX, double mouseY, int button) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new MouseClickedEvent((float) mouseX, (float) mouseY, button));
		}
		
		return super.mouseClicked(mouseX, mouseY, button);
	}
	
	@Override
	public boolean mouseReleased(double mouseX, double mouseY, int button) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new MouseReleasedEvent((float) mouseX, (float) mouseY, button));
		}
		
		return super.mouseReleased(mouseX, mouseY, button);
	}
	
	@Override
	public boolean mouseDragged(double mouseX, double mouseY, int button, double deltaX, double deltaY) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new MouseDraggedEvent((float) mouseX, (float) mouseY, button, deltaX, deltaY));
		}
		
		return super.mouseDragged(mouseX, mouseY, button, deltaX, deltaY);
	}
	
	@Override
	public void mouseMoved(double mouseX, double mouseY) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new MouseMovedEvent((float) mouseX, (float) mouseY));
		}
		
		super.mouseMoved(mouseX, mouseY);
	}
	
	@Override
	public boolean mouseScrolled(double mouseX, double mouseY, double amount) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new MouseScrolledEvent((float) mouseX, (float) mouseX, amount));
		}
		
		return super.mouseScrolled(mouseX, mouseY, amount);
	}
	
	@Override
	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new KeyPressedEvent(keyCode, scanCode, modifiers));
		}
		
		return super.keyPressed(keyCode, scanCode, modifiers);
	}
	
	@Override
	public boolean keyReleased(int keyCode, int scanCode, int modifiers) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new KeyReleasedEvent(keyCode, scanCode, modifiers));
		}
		
		return super.keyReleased(keyCode, scanCode, modifiers);
	}
	
	@Override
	public boolean charTyped(char chr, int modifiers) {
		for (var child : handler.getChildren()) {
			child.dispatchEvent(new CharacterTypedEvent(chr, modifiers));
		}
		
		return super.charTyped(chr, modifiers);
	}
	
	@Override
	public void render(MatrixStack matrices, int mouseX, int mouseY, float delta) {
		super.renderBackground(matrices);
		
		var client = InstanceUtil.getClient();
		
		var provider = client.getBufferBuilders().getEffectVertexConsumers();
		
		for (var child : handler.getChildren()) {
			if (!child.isHidden()) {
				child.draw(matrices, provider, delta);
			}
		}
		
		var mousePos = PositionUtil.getMousePosition();
		
		Widget minChild = null;
		float minDist = Float.MAX_VALUE;
		
		for (var child : handler.getAllChildren()) {
			if (!child.isHidden() && child.isFocused()) {
				if (minChild == null || minDist == Float.MAX_VALUE || minChild.getPosition().distanceTo(mousePos) > child.getPosition().distanceTo(mousePos)) {
					minChild = child;
					minDist = child.getPosition().distanceTo(mousePos);
				}
			}
		}
		
		if (minChild != null) {
			renderTooltip(matrices, (List<Text>) minChild.getTooltips(), mouseX, mouseY);
		}
		
		provider.draw();
		
		super.render(matrices, mouseX, mouseY, delta);
		
		super.drawMouseoverTooltip(matrices, mouseX, mouseY);
	}
}

File: src/main/kotlin/dev/vini2003/hammer/client/screen/BaseHandledScreen.kt
package dev.vini2003.hammer.client.screen

import dev.vini2003.hammer.client.util.Instances
import dev.vini2003.hammer.common.screen.handler.BaseScreenHandler
import dev.vini2003.hammer.common.util.Networks
import dev.vini2003.hammer.common.widget.Widget
import net.minecraft.client.gui.screen.ingame.HandledScreen
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.text.Text

open class BaseHandledScreen<T : BaseScreenHandler>(handler: BaseScreenHandler, inventory: PlayerInventory, title: Text) : HandledScreen<T>(handler as T, inventory, title) {
	override fun init() {
		handler.widgets.clear()
		handler.slots.clear()
		
		backgroundWidth = width
		backgroundHeight = height
		
		super.init()
		
		handler.initialize(width, height)
		handler.onLayoutChanged()
		
		handler.allWidgets.forEach(Widget::onLayoutChanged)
		
		Networks.toServer(Networks.Initialize, Networks.ofInitialize(handler.syncId, width, height))
	}

	override fun drawBackground(matrices: MatrixStack?, delta: Float, mouseX: Int, mouseY: Int) {

	}

	override fun isClickOutsideBounds(mouseX: Double, mouseY: Double, left: Int, top: Int, button: Int): Boolean {
		return handler.allWidgets.none { widget -> widget.isWithin(mouseX.toFloat(), mouseY.toFloat()) }
	}

	override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
		handler.widgets.forEach {
			it.onMouseClicked(mouseX.toFloat(), mouseY.toFloat(), button)
		}

		return super.mouseClicked(mouseX, mouseY, button)
	}

	override fun mouseReleased(mouseX: Double, mouseY: Double, button: Int): Boolean {
		handler.widgets.forEach {
			it.onMouseReleased(mouseX.toFloat(), mouseY.toFloat(), button)
		}

		return super.mouseReleased(mouseX, mouseY, button)
	}

	override fun mouseDragged(mouseX: Double, mouseY: Double, button: Int, deltaX: Double, deltaY: Double): Boolean {
		handler.widgets.forEach {
			it.onMouseDragged(mouseX.toFloat(), mouseY.toFloat(), button, deltaX, deltaY)
		}

		return super.mouseDragged(mouseX, mouseY, button, deltaX, deltaY)
	}

	override fun mouseMoved(mouseX: Double, mouseY: Double) {
		handler.widgets.forEach {
			it.onMouseMoved(mouseX.toFloat(), mouseY.toFloat())
		}

		super.mouseMoved(mouseX, mouseY)
	}

	override fun mouseScrolled(mouseX: Double, mouseY: Double, deltaY: Double): Boolean {
		handler.widgets.forEach {
			it.onMouseScrolled(mouseX.toFloat(), mouseY.toFloat(), deltaY)
		}

		return super.mouseScrolled(mouseX, mouseY, deltaY)
	}

	override fun keyPressed(keyCode: Int, scanCode: Int, keyModifiers: Int): Boolean {
		handler.widgets.forEach {
			it.onKeyPressed(keyCode, scanCode, keyModifiers)
		}

		return super.keyPressed(keyCode, scanCode, keyModifiers)
	}

	override fun keyReleased(keyCode: Int, scanCode: Int, keyModifiers: Int): Boolean {
		handler.widgets.forEach {
			it.onKeyReleased(keyCode, scanCode, keyModifiers)
		}

		return super.keyReleased(keyCode, scanCode, keyModifiers)
	}

	override fun charTyped(character: Char, keyCode: Int): Boolean {
		handler.widgets.forEach {
			it.onCharacterTyped(character, keyCode)
		}

		return super.charTyped(character, keyCode)
	}

	override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
		super.renderBackground(matrices)

		val provider: VertexConsumerProvider.Immediate = Instances.client.bufferBuilders.effectVertexConsumers

		handler.widgets.asSequence().filterNot(Widget::hidden).forEach {
			it.drawWidget(matrices, provider)
		}
		
		handler.allWidgets.asSequence().filterNot(Widget::hidden).filter(Widget::focused).forEach {
			renderTooltip(matrices, it.getTooltip(), mouseX, mouseY)
		}

		provider.draw()

		super.render(matrices, mouseX, mouseY, delta)

		super.drawMouseoverTooltip(matrices, mouseX, mouseY)
	}
}